### DISCUSSION TITLE
Query if file exists, or more generally, query the file system

### DISCUSSION THREAD

**Discussion by rbeeli** [ClickHouse/ClickHouse#62775] [Q&A]
It is easily possible to query local files relative to the `user_files_path` setting, e.g.
```sql
select * from file('path_to_file.parquet')
```

What I have not managed to figure out is if it is possible to check if a file actually exists without throwing an error? Something like:

```sql
select exists file('path_to_file.parquet')
```
or similar.

An extension of this functionality would be to just being able to query the file system itself using glob patterns, which is already implicitly possible, e.g.
```sql
select * from file('folder/*.parquet')
```

Such a feature would also be useful for object storage providers like S3 and Azure, not just local file system.

**Comment by yiyang-shao** [comment]
Hi! Could you check if you can use [this setting](https://clickhouse.com/docs/en/operations/settings/settings#engine-file-empty_if-not-exists) as a workaround?

**Reply by rbeeli** [reply]
I don't see a difference when setting this option, still gives the same `Code: 400. DB::ErrnoException: Cannot stat file` error. Any other solutions, or is this a missing feature?
Thanks.

**Reply by UnamedRus** [reply]
```
select DISTINCT _file from file('folder/*.parquet', 'One')
```

**Reply by rbeeli** [reply]
Hi @UnamedRus,

Thanks for the suggestion!

This seems like the best solution so far, though it still does a list operation of all directory files, which is not optimal in terms of performance. Would be nice to have a dedicated way to do this though, a lot of magic with the `_file` column and `'One'` parameter.

What I have now:
`select 1 exists from file('FOLDER_PATH/*.parquet', 'One') where _file = 'MY_FILE.parquet'`
